pkg/distributor/distributor_test.go

```diff
@@ -2773,6 +2772,212 @@ func TestDistributor_LabelValuesCardinality_Concurrency(t *testing.T) {
 	})
 }
 
+func TestHaDedupeMiddleware(t *testing.T) {
+	ctxWithUser := user.InjectOrgID(context.Background(), "user")
+	const replica1 = "replicaA"
+	const replica2 = "replicaB"
+	const cluster1 = "clusterA"
+	const cluster2 = "clusterB"
+
+	type testCase struct {
+		name              string
+		ctx               context.Context
+		enableHaTracker   bool
+		acceptHaSamples   bool
+		reqs              []*mimirpb.WriteRequest
+		expectedReqs      []*mimirpb.WriteRequest
+		expectedNextCalls int
+		expectErrs        []int
+	}
+	testCases := []testCase{
+		{
+			name:              "no changes on empty request",
+			ctx:               ctxWithUser,
+			enableHaTracker:   true,
+			acceptHaSamples:   true,
+			reqs:              []*mimirpb.WriteRequest{{}},
+			expectedReqs:      []*mimirpb.WriteRequest{{}},
+			expectedNextCalls: 1,
+			expectErrs:        []int{0},
+		}, {
+			name:              "no changes if accept HA samples is false",
+			ctx:               ctxWithUser,
+			enableHaTracker:   true,
+			acceptHaSamples:   false,
+			reqs:              []*mimirpb.WriteRequest{makeWriteRequestForLabelSetGen(5, labelSetGenWithReplicaAndCluster(replica1, cluster1))},
+			expectedReqs:      []*mimirpb.WriteRequest{makeWriteRequestForLabelSetGen(5, labelSetGenWithReplicaAndCluster(replica1, cluster1))},
+			expectedNextCalls: 1,
+			expectErrs:        []int{0},
+		}, {
+			name:              "remove replica label with HA tracker disabled",
```

Comment:
Do we remove replica label when HA tracker is disabled? 🤔

Reply:
yes before this change we removed it when the HA tracker was disabled, so I kept that behavior.

`checkSample` calls `d.HATracker.checkReplica`: https://github.com/grafana/mimir/blob/1837e074221846f6b30bd09595987aa48ba0f4e3/pkg/distributor/distributor.go#L549

`checkReplica` returns `nil` when the tracker is disabled: https://github.com/grafana/mimir/blob/0ed77009e28df5227f06924b9133bbe58b7a8867/pkg/distributor/ha_tracker.go#L415

if `checkReplica` returns `nil` then `checkSample` returns `true`: https://github.com/grafana/mimir/blob/1837e074221846f6b30bd09595987aa48ba0f4e3/pkg/distributor/distributor.go#L555

so `removeReplica` will be `true` https://github.com/grafana/mimir/blob/1837e074221846f6b30bd09595987aa48ba0f4e3/pkg/distributor/distributor.go#L766